packages/next/src/server/route-modules/app-route/module.ts

```diff
@@ -350,6 +350,7 @@ export class AppRouteRouteModule extends RouteModule<
           const prospectiveRoutePrerenderStore: PrerenderStore =
             (prerenderStore = {
               type: 'prerender',
+              phase: 'action',
```

Comment:
@sebmarkbage i'm not sure how to handle the case of prerendering route modules -- giving them `phase: 'action'` feels a bit strange, because they definitely shouldn't be able to do anything mutable. but i guess we'd block that by other means (i.e. dynamic APIs throwing during a prerender) so maybe it's fine?

Reply:
I think what we want is prerenders are run in a sort of `'render'` mode but if you do anything that would have required `'action'` phase we just treat it as dynamic. This is tricky b/c right now trying to set a cookie in render mode is an error. So maybe we run these prerenders in action mode and just rely on dynamic tracking to kick us out of prerendering since all of the things that require action mode also trigger dynamic behavior?

Reply:
Hmm... I wonder if statically optimizable GET handlers need to be annotated at build time to run in `'render'` phase always even when triggered dynamically. And all other GET handlers need to be annotated at build time to run in `'action'` phase. It's odd that a route handler could not set cookies during prerender but then if you happened to hit the route handler dynamically it would allow cookies to be set, but if the handler became static again it would again disallow handlers.

By testing the route for whether it is statically optimizable at build time we can discover whether the handler adheres to render phase semantics. If it does we should run it with those semantics all the time even if we happen to trigger it dynamically. And if we can't statically optimize it we assume it has action semantics and always consistently run it with those.

Not sure if routes can even be hit dynamically when they are statically optimized so maybe this is a moot point but the invariant I think we need to hold to is we don't use one phase during prerenders and another during dynamic renders for the same handler